Add WC Product CSV Exporter limit filter

AdamQuadmon opened the pull request:
with tons of products is nice to have an way to change the limit for the exporter (now is set to 50)

```php
add_filter( 'woocommerce_product_export_get_limit', function ( $limit, $export_type ) {
  if( 'product' === $export_type) {
    $limit = 100;
  }

  return $limit
}, 10, 2);

AdamQuadmon committed:
Add WC Product CSV Exporter limit filter

with tons of products is nice to have an way to change the limit for the exporter (now is set to 50)

'''php
function wc_exporter_set_limit($limit) {
  return 1000;
}
add_filter( 'woocommerce_product_export_product_default_limit', 'wc_exporter_set_limit');
'''

WPprodigy commented:
Filtering there could cause some problems. For example, this function will be messed up https://github.com/woocommerce/woocommerce/blob/f0846af5332e347a51df88a14fce283225ac940c/includes/export/abstract-wc-csv-batch-exporter.php#L138

Should add the filter here I think: https://github.com/woocommerce/woocommerce/blob/e6e61b084a1d10d8eb1081fd59ed887a079e70a1/includes/export/abstract-wc-csv-exporter.php#L298

Also, don't forget the code standards :). Namely spacing https://make.wordpress.org/core/handbook/best-practices/coding-standards/php/#space-usage

AdamQuadmon commented:
Hi @WPprodigy, 
the filter proposed add a default value for `limit` in the constructor via setter

`get_limit()` will return the set value, and `get_total_exported()` works just fine.

having the filter in the `get_limit()` function will screw up plugins that wants to change via setter the value.

but I get your point, maybe is better to set the default value in the `WC_CSV_Batch_Exporter` constructor instead of the `WC_Product_CSV_Exporter` ones.

If so I'll do the changes and the spaces fix

WPprodigy commented:
Right, adding it there makes the filter useful for both WC_CSV_Batch_Exporter and WC_Product_CSV_Exporter was my thought process.

Could pass in `$this`, or maybe just `$this->export_type` as a filter argument as well to add more extensibility for export types in the future.

AdamQuadmon committed:
Rename exporter filter to `woocommerce_product_export_init`

moving it in the abstract constructor and let it be more generic now is possible to do:
```
add_filter( 'woocommerce_product_export_init', function ( $exporter ) {
  if( 'product' === $exporter->export_type ) {
    $exporter->set_limit(100);
  }
});

AdamQuadmon committed:
fix not needed double quotes

AdamQuadmon commented:
It looks fine, the travis failure is due to previous code style errors in the file

the only issue here is that having the filter in the abstract constructor does not let you check for the export_type

claudiosanches commented on includes/export/abstract-wc-csv-batch-exporter.php:
```diff
@@ -46,6 +46,7 @@ public function __construct() {
 		$upload_dir         = wp_upload_dir();
 		$this->file         = trailingslashit( $upload_dir['basedir'] ) . $this->get_filename();
 		$this->column_names = $this->get_default_column_names();
+		apply_filters( 'woocommerce_product_export_init', $this );
```
This is filter is not applied in anything here.
Not sure what you indented to do, better you test first.

AdamQuadmon replied:
```php
add_filter( 'woocommerce_product_export_init', function ( $exporter ) {
  $exporter->set_limit(100);
});
```

claudiosanches replied:
An filter need to always return a value. Looks like you want to declare an action...

AdamQuadmon replied:
right, didn't thought about that, I'm new to WP.
anyway we changed filter position and now it's returning.

WPprodigy commented:
Still think the best place to filter is in `get_limit` on the base class. Filtering on the `get` method in general seems to be the common practice. 

> having the filter in the get_limit() function will screw up plugins that wants to change via setter the value.

Not really - the filter will just override the plugin's setting. Said plugin should still work fine because they should still be using the get method.

If the filter is used without any conditions, then the user should expect it to alter all plugins that extend the class. That's why passing `$this->export_type` as a filter argument makes it more flexible.

AdamQuadmon committed:
change filter name and position `woocommerce_product_export_get_limit`

AdamQuadmon commented:
@WPprodigy 

```php
add_filter( 'woocommerce_product_export_get_limit', function ( $limit, $export_type ) {
  if( 'product' === $export_type) {
    $limit = 100;
  }

  return $limit
}, 10, 2);
```

claudiosanches commented:
@AdamQuadmon please explain to my this:

> with tons of products is nice to have an way to change the limit for the exporter (now is set to 50)

Still all products will be imported. But are importer 50 at once, after more 50, and so on by AJAX.
This is just to not overload your machine importing all products at once.

AdamQuadmon commented:
@claudiosanches you are right, a dev should test the machine can handle it.

my environment is a vagrant machine that emulates the production one

currently I'm exporting products with 10 attributes with limit set to 1000 on a 5$/mo DO VPS

claudiosanches commented:
@AdamQuadmon ok. You haven't replied my question yet... But.

> my environment is a vagrant machine that emulates the production one

Emulate the production one with all user traffic? People browsing, placing orders, getting downloadable products and so on? 

As I said, the limit is just for the queue, but the file will be generated anyway with all products at the end. You can change the queue to 1000, still your server may run out of resources and stop exporting, or even worse like have corrupt exported data, at least with less products are sure that the exportation process will finish without errors.

I appreciate your contribution, if you can send a Pull Request that really allows to change or filter this limit, I'll be happy to merge.

claudiosanches commented on includes/export/abstract-wc-csv-exporter.php:
```diff
@@ -296,7 +296,7 @@ protected function export_row( $row_data, $key, $buffer ) {
 	 */
 	public function get_limit() {
-		return $this->limit;
+		return apply_filters( "woocommerce_product_export_get_limit", $this->limit, $this->export_type);
```
Now should works.
But need some coding style fixes.

- [ ] Need to replace double quotes to single quotes.
- [ ] Missing an empty space after `$this->export_type`.

claudiosanches replied:
In fact, better change to:

```php
return apply_filters( "woocommerce_{$this->export_type}_export_batch_limit", $this->limit, $this );
```

claudiosanches commented on includes/export/abstract-wc-csv-exporter.php:
```diff
@@ -296,7 +296,7 @@ protected function export_row( $row_data, $key, $buffer ) {
 	 */
 	public function get_limit() {
-		return $this->limit;
+		return apply_filters( "woocommerce_product_export_get_limit", $this->limit, $this->export_type);
```
Better change to:

```php
return apply_filters( "woocommerce_{$this->export_type}_export_batch_limit", $this->limit, $this );
```

Better filter naming, since this that class may or not be for products, and this fix the coding style issues too.

AdamQuadmon committed:
change filter name

claudiosanches commented:
@AdamQuadmon thanks for your contribution.

AdamQuadmon commented:
@claudiosanches happy to contribute!
